gdal/frmts/pdf/pdfcreatecopy.cpp

```diff
@@ -2164,6 +2165,114 @@ static void DrawGeometry(VSILFILE* fp, OGRGeometryH hGeom, double adfMatrix[4],
+    };
+
+    // Character widths of Times-Roman, Win-1252 characters 32 to 255
+    // Times bold, oblique and bold oblique have their own widths,
+    // but for now we will put up with these widths on all Times variants
+    static const GUInt16 anTimesRomanCharWidths[] = {
+        512, 682, 836, 1024,1024,1706,1593,369, 682, 682, 1024,1155,512, 682, 512, 569,
+        1024,1024,1024,1024,1024,1024,1024,1024,1024,1024,569, 569, 1155,1155,1155,909,
+        1886,1479,1366,1366,1479,1251,1139,1479,1479,682, 797, 1479,1251,1821,1479,1479,
+        1139,1479,1366,1139,1251,1479,1479,1933,1479,1479,1251,682, 569, 682, 961, 1024,
+        682, 909, 1024,909, 1024,909, 682, 1024,1024,569, 569, 1024,569, 1593,1024,1024,
+        1024,1024,682, 797, 569, 1024,1024,1479,1024,1024,909, 983, 410, 983, 1108,0,
+        1024,2048,682, 1024,909, 2048,1024,1024,682 ,2048,1139,682 ,1821,2048,1251,2048,
+        2048,682, 682, 909, 909, 717 ,1024,2048,682 ,2007,797, 682 ,1479,2048,909, 1479,
+        512, 682, 1024,1024,1024,1024,410, 1024,682, 1556,565, 1024,1155,682, 1556,1024,
+        819, 1124,614, 614, 682, 1180,928, 682, 682, 614, 635, 1024,1536,1536,1536,909,
+        1479,1479,1479,1479,1479,1479,1821,1366,1251,1251,1251,1251,682, 682, 682, 682,
+        1479,1479,1479,1479,1479,1479,1479,1155,1479,1479,1479,1479,1479,1479,1139,1024,
+        909, 909, 909, 909, 909, 909, 1366,909, 909, 909, 909, 909, 569, 569, 569, 569,
+        1024,1024,1024,1024,1024,1024,1024,1124,1024,1024,1024,1024,1024,1024,1024,1024
+    };
+
+    const GUInt16* panCharacterWidths = nullptr;
+
+    if( STARTS_WITH_CI( osFont, "times" ) ||
+        osFont.find( "Serif", 0 ) != std::string::npos )
+    {
+        if( bBold && bItalic )
+            osFont = "Times-BoldItalic";
+        else if( bBold )
+            osFont = "Times-Bold";
+        else if( bItalic )
+            osFont = "Times-Italic";
+        else
+            osFont = "Times-Roman";
+
+        panCharacterWidths = anTimesRomanCharWidths;
+        dfHeight = dfSize * 1356.0 / 2048;
+    }
+    else if( STARTS_WITH_CI( osFont, "courier" ) ||
+        osFont.find( "Mono", 0 ) != std::string::npos )
+    {
+        if( bBold && bItalic )
+            osFont = "Courier-BoldOblique";
+        else if( bBold )
+            osFont = "Courier-Bold";
+        else if( bItalic )
+            osFont = "Courier-Oblique";
+        else
+            osFont = "Courier";
+
+        dfHeight = dfSize * 1170.0 / 2048;
+    }
+    else
+    {
+        if( bBold && bItalic )
+            osFont = "Helvetica-BoldOblique";
+        else if( bBold )
+            osFont = "Helvetica-Bold";
+        else if( bItalic )
+            osFont = "Helvetica-Oblique";
+        else
+            osFont = "Helvetica";
+
+        panCharacterWidths = anHelveticaCharWidths;
+        dfHeight = dfSize * 1467.0 / 2048;
+    }
+
+    dfWidth = 0.0;
+    for( const char& ch : osText )
+    {
+        const int nCh = static_cast<int>( ch );
```

Comment:
shouldn't that be cast to unsigned char instead ? Otherwise I suspect any character > 127 will be seen as a negative value